Consolidating "vignette" & "workflow" layers into "pipeline" layer (#202)
getwilds/wilds-wdl-library: pipelines/README.md

# WILDS WDL Pipelines

WILDS WDL pipelines are complete analysis workflows that combine multiple modules to create functional bioinformatics pipelines. These range from compact educational examples to comprehensive production-ready analyses, all sharing a common structure and design philosophy.

## Pipeline Philosophy

The WILDS pipeline system is designed around several key principles:

### **Module Integration**
Each pipeline demonstrates how to:
- Import and combine multiple WILDS WDL modules effectively
- Pass data seamlessly between different tools
- Coordinate resource allocation across modules
- Handle complex data structures and dependencies

### **Real-World Patterns**
Pipelines focus on workflows that are:
- Commonly used in bioinformatics research
- Representative of standard analysis patterns
- Suitable for production use with real datasets
- Extensible for specific research needs

### **Educational Value**
Every pipeline serves as:
- A working example of best practices
- Documentation of integration patterns
- A template for similar workflows
- A testing framework for module compatibility

## Complexity Levels

> Pipelines vary in scope and complexity. Each pipeline's README documents its complexity level to help you choose the right starting point:

Reviewer: ```suggestion
Pipelines vary in scope and complexity. Each pipeline and its complexity level is documented below to help you choose the right starting point:
```
The pipeline READMEs don't have the complexity levels, but maybe it's ok to keep all that info in this README?

| Level | Modules | Typical Runtime | Description |
|-------|---------|-----------------|-------------|
| **Basic** | 2-3 | < 30 minutes | Simple integrations ideal for learning |
| **Intermediate** | 4-6 | 1-4 hours | Multi-step analyses for common use cases |
| **Advanced** | 10+ | > 4 hours | Comprehensive production pipelines |

## Pipeline Structure

Each pipeline directory contains the following standard components:

```
pipelines/pipeline-name/
├── pipeline-name.wdl    # Main workflow importing multiple modules
├── testrun.wdl          # Zero-configuration test workflow
├── inputs.json          # Example inputs for the complete pipeline
└── README.md            # Pipeline-specific documentation
```

### **Required Components**

- **WDL File**: Main workflow that imports and combines modules
- **Test Workflow**: `testrun.wdl` - Zero-configuration demonstration workflow with hardcoded test inputs
- **Inputs JSON**: Example input configurations to use as a starting point for real data

> - **README**: Detailed documentation with usage examples and integration patterns

Reviewer: ```suggestion
```
This is repeating the diagram above

### **Optional Components**

- **Options JSON**: Workflow execution configuration for different executors

> - **Additional Input Files**: Alternative configurations for different use cases

Reviewer: ```suggestion
- **Options JSON**: Workflow execution configuration for different WDL engines
- **Additional Input Files**: Alternative inputs for different use cases
```
Just some suggestions

sprocket run ww-sra-star.wdl inputs.json
```

This works because all pipelines import modules using GitHub URLs, so your WDL executor fetches dependencies automatically.

### **Testing and Demonstration**

Pipelines also include zero-configuration test workflows for quick demonstrations:

```bash
# Download and run a test workflow (no inputs needed)
curl -O https://raw.githubusercontent.com/getwilds/wilds-wdl-library/main/pipelines/ww-sra-star/testrun.wdl
sprocket run testrun.wdl
```

If you have the repository cloned:

```bash
# Navigate to pipeline directory
cd pipelines/ww-sra-star

# Run test workflow with your preferred executor (no inputs needed)
sprocket run testrun.wdl
java -jar cromwell.jar run testrun.wdl
miniwdl run testrun.wdl
```

### **As Complete Workflows**

Pipelines can be executed as standalone workflows with your own data:

```bash
# Navigate to pipeline directory
cd pipelines/ww-sra-star

# Run with your preferred executor using the example inputs.json as a starting point
miniwdl run ww-sra-star.wdl -i inputs.json
java -jar cromwell.jar run ww-sra-star.wdl --inputs inputs.json
sprocket run ww-sra-star.wdl inputs.json

> ```

Reviewer: ```suggestion
```bash
# Navigate to pipeline directory
cd pipelines/ww-sra-star

# Run with your preferred executor using the example inputs.json as a starting point:

# Sprocket
sprocket run ww-sra-star.wdl inputs.json

# Cromwell
java -jar cromwell.jar run ww-sra-star.wdl --inputs inputs.json

# miniWDL
miniwdl run ww-sra-star.wdl -i inputs.json
```
```

### **As Templates**

Pipelines serve as starting points for custom workflows:
- Copy and modify for specific research needs
- Add additional modules or analysis steps
- Adjust parameters for different datasets
- Extend with quality control or visualization components

> 

Reviewer: ```suggestion

```
I think this is covered at the very top

Author: I do think it's lightly covered at the top, but I kind of like reiterating it here as a major use case. Going to leave it in if that's alright.

## Testing and Validation

### **Automated Testing**
All pipelines are automatically tested through GitHub Actions:
- **Multi-executor testing**: Cromwell, miniWDL, and Sprocket
- **End-to-end validation**: Complete pipeline testing with real data

> - **Integration verification**: Ensures modules work together correctly

Reviewer: ```suggestion
```
I think this is implied in the line before

### **Test Data**
- Small but realistic datasets for efficient testing
- Publicly available reference data

> - Automated download and setup during CI

Reviewer: ```suggestion
- Automatically downloaded during continuous integration
```

## Integration Patterns

Pipelines demonstrate key WDL patterns:
- **Module Composition**: How to effectively combine different tools
- **Data Flow Management**: Passing complex outputs between modules
- **Resource Coordination**: Balancing compute requirements across steps

## Contributing

### **Adding New Pipelines**
1. Fork the repository
2. Create a new pipeline directory following the standard structure
3. Import existing modules (prefer existing modules over creating new tasks)
4. Document the complexity level in your README
5. Add comprehensive documentation and test inputs
6. Submit a pull request demonstrating the integration pattern

### **Improving Existing Pipelines**
- Performance optimizations and resource tuning
- Enhanced documentation and usage examples
- Additional input configurations for different use cases
- Integration of new modules as they become available

## Support and Feedback

For questions about pipelines or to report issues:
- Open an issue in the [WILDS WDL Library repository](https://github.com/getwilds/wilds-wdl-library/issues)
- Contact the Fred Hutch Data Science Lab at wilds@fredhutch.org

> - See the [WILDS Contributor Guide](https://getwilds.org/guide/) for detailed guidelines

Reviewer: ```suggestion
- See the [WILDS Contributor Guide](https://github.com/getwilds/wilds-wdl-library/blob/main/.github/CONTRIBUTING.md) for detailed guidelines
```
I would point to our more current GitHub contributing page
